### PR TITLE
SubnetworkRecCell: resolve beams of seq_len

### LINKED ISSUE #225: SubnetworkRecCell: size_placeholder of outputs not resolved
I have the following configuration in my recurrent unit:

Choice layers C1 and C2, end layer E
C2 depends on C1, but not on E
E depends on C1

C2 is the last choice layer, thus defines the beam to which all outputs are resolved to (via `_opt_search_resolve()`). C1 is resolved to C2, this works.
However, the size_placeholder of all outputs is set to `seq_len` (in `get_loop_acc_layer()`), which itself comes from the length information of the end layer. In my case, the end layer has beams according to C1, not C2. Resolving these beams is currently missing, giving a shape mismatch in case of different beam sizes or, even worse, outputs of wrong length in case of same beam sizes.

I'm trying to figure out where to do this, but maybe you, @albertz, have an idea right away.

### PR DESCRIPTION
**Pull request by patrick-wilken**
This fixes  #225

Two things to discuss:
1. Resolving beams in `body()` is now pretty inconsistent (`translate_to_this_search_beam`, `SelectSearchSourcesLayer.transform_func` and `select_src_beams`). I think we should refactor the code, such that there is one utility function, which translates beams taking a sequence of search choices, and `SelectSearchSourcesLayer` should use this function.

2. It is a bit vague how exactly the latest search choice in the recurrent unit should be determined. I now calculate it globally, comparing all search choices of `needed_outputs`. Is there any downside to this? It should give the same result as the first "prev" layer in the choice sequence and it should be enough to do it only once, but there is simply no obvious layer to start the choice sequence from ("output" can be outside the loop, "end" doesn't always exists). 
And for the initial shape of `dyn_seq_len` I use the "output" layer template, which assumes it to have the resolved output beam size... Maybe we should also use the latest choice here, but this would require getting it from the templates.

### CODE REVIEW

File: TFNetworkRecLayer.py [hunk 1]
```diff
@@ -2214,9 +2268,10 @@ def cond(i, net_vars, acc_tas, seq_len_info=None):
       [tf.TensorShape(None) for _ in init_acc_tas])
     if not have_known_seq_len:
       # See body().
-      out_batch_dim = self.layer_data_templates["end"].get_batch_dim()
+      end_batch_dim = self.layer_data_templates["end"].get_batch_dim()
+      out_batch_dim = self.layer_data_templates["output"].get_batch_dim()
       init_seq_len_info = (
-        constant_with_shape(False, shape=[out_batch_dim], name="initial_end_flag"),
+        constant_with_shape(False, shape=[end_batch_dim], name="initial_end_flag"),
         constant_with_shape(0, shape=[out_batch_dim], name="initial_seq_len"))
```
Comment (albertz) [inline TFNetworkRecLayer.py]: I don't think it is a good idea to mix it. As said, the end layer should just be resolved automatically to the latest choice. And the same for the "output" layer.

File: TFNetworkRecLayer.py [hunk 2]
```diff
@@ -2135,13 +2178,24 @@ def maybe_transform(layer):
...
+              # Translate the beams of end_flags to the latest choices in the net.
+              last_choices = self._latest_choice_layer.get_search_choices()
+              resolved_end_layer = last_choices.translate_to_this_search_beam(cur_end_layer)
+              end_flag_resolved = resolved_end_layer.transform_func(
+                cur_end_layer.output.placeholder if rec_layer.include_eos else end_flag)
             with tf.name_scope("dyn_seq_len"):
-              dyn_seq_len = cur_end_layer.transform_func(dyn_seq_len)
+              # Translate beams of dyn_seq_len from previous to current frame (going through all choices)
+              for choice_layer in reversed(self._choice_seq_in_frame):
+                dyn_seq_len = select_src_beams(dyn_seq_len, src_beams=choice_layer.search_choices.src_beams)
```
Comment (albertz) [inline TFNetworkRecLayer.py]: Don't call `select_src_beams` here, and esp do not replicate the logic about unrolling over the choice seq. Just use the existing logic (e.g. call `transform_func`).
In any case, as said, seq len and end flag, both should be resolved in the same way.

**Comment by albertz** [comment]
> Resolving beams in body() is now pretty inconsistent (translate_to_this_search_beam, SelectSearchSourcesLayer.transform_func and select_src_beams). I think we should refactor the code, such that there is one utility function, which translates beams taking a sequence of search choices, and SelectSearchSourcesLayer should use this function.

But there is only one such place. This is what the `SelectSearchSourcesLayer` does. You basically never should call `select_src_beams` explicitly, this should be all via `SelectSearchSourcesLayer`. `SelectSearchSourcesLayer.transform_func` was just an export of that functionality.

**Comment by albertz** [comment]
> It is a bit vague how exactly the latest search choice in the recurrent unit should be determined.

Why? This is perfectly defined currently. See `_opt_search_resolve`, where we define `latest_layer_choice`.

> I now calculate it globally, comparing all search choices of needed_outputs. 

This would be wrong. E.g. if there are two separate choices C1 and C2, which are not dependent at all on each other. Then there is not a single global search choice. Every single layer can basically have it's own latest layer choice.

**Comment by albertz** [comment]
Btw, in any case, irrelevant how you fix it (#225), you should also add a simple test case here which fails before the fix, but works after the fix.

File: TFNetworkRecLayer.py (hunk 2 above)
Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: Yeah, I was aware of this, but the alternative here would be to get the "prev:_latest_choice_layer", transform it to the current frame with `translate_to_this_search_beam` and then never use the output, but only use `transform_function` to transform `dyn_seq_len`. This didn't feel right either... Or is there a better way?

**Comment by patrick-wilken** [comment]
> This would be wrong. E.g. if there are two separate choices C1 and C2, which are not dependent at all on each other. Then there is not a single global search choice. Every single layer can basically have it's own latest layer choice.

Oh wow, true, but this makes it complicated. So then there is no such thing as a single beam of seq_lens which can be used to set size_placeholder of the outputs as we currently do it in `_construct_output_layers_moved_out`...

**Comment by albertz** [comment]
Even with multiple independent choices, there is still a single ending condition (must be), which depends (or might depend) on one of the choices. And seq len relates to the end flag.

**Comment by patrick-wilken** [comment]
Rewrote basically the whole PR and added test cases.

File: TFNetworkRecLayer.py [hunk 3]
```diff
@@ -2350,33 +2351,34 @@ def cond(i, net_vars, acc_tas, seq_len_info=None):
...
     """
     import os
     from TFUtil import nd_indices, assert_min_tf_version, expand_dims_unbroadcast, get_valid_scope_name_from_str
     from TFUtil import get_shape_dim
-    from TFNetwork import LayerNotFound
     rec_layer = self.parent_rec_layer
     try:
       layer = self.net.get_layer(layer_name)
     except LayerNotFound:  # layer is not inside loop
       return acc_ta, None
```
Comment (albertz) [inline TFNetworkRecLayer.py]: This is wrong now.

**Comment by patrick-wilken** [comment]
Hmm, somehow I cannot add inline comments at the moment...

What I do now is keep the sequence lengths on the end layer beam and only translate the beam afterwards to the different outputs. This allows different beams for different outputs.

The part I added is in part copy pasted from `search_resolve_body`, maybe we should avoid this code duplication.

The test `test_attention_two_dependent_targets()` fails before this commit. The other test fails before 0e23bcd.

File: TFNetworkRecLayer.py [hunk 4]
```diff
@@ -2834,19 +2879,20 @@ def get_loop_acc_layer(name):
...
-          layer_name=name, acc_ta=acc_ta, final_net_vars=final_net_vars)
+        acc_ta, search_choices, resolved_seq_len = self._opt_search_resolve(
+          layer_name=name, acc_ta=acc_ta, final_net_vars=final_net_vars, seq_len=seq_len)
         output = self.layer_data_templates[name].output.copy_template_adding_time_dim(time_dim_axis=0)
         output.beam = search_choices.get_beam_info() if search_choices else None
+        max_len = tf.reduce_max(resolved_seq_len) if resolved_seq_len is not None else None
         # We should have accumulated it.
         output.placeholder = tensor_array_stack(acc_ta, stop=max_len)  # e.g. (time,batch,dim)
-        output.size_placeholder = {0: seq_len}
+        output.size_placeholder = {0: resolved_seq_len}
```
Comment (albertz) [inline TFNetworkRecLayer.py]: Can it be `None`? (You check for that above.) If yes, than this is wrong. If no, then the code above does not make sense.

File: TFNetworkRecLayer.py [hunk 5]
```diff
@@ -2431,11 +2433,56 @@ def get_choice_seq(choice_base):
...
+
+      # Separate batch and beam dims
+      seq_len_beam_size = end_layer.output.beam.beam_size if end_layer else rec_layer.output.beam.beam_size
+      seq_len = tf.reshape(
+        seq_len, [batch_dim, seq_len_beam_size],
+        name="split_batch_beam")
+
+      if end_layer:
+        end_layer_choice = self.net.get_search_choices(src=end_layer)
+        assert end_layer_choice in choice_seq_in_frame,\
```
Comment (albertz) [inline TFNetworkRecLayer.py]: Don't use `\`. Rather `assert ..., (\n...)`

File: TFNetworkRecLayer.py [hunk 6]
```diff
@@ -8,6 +8,7 @@
 import tensorflow as tf
 import typing
 from tensorflow.python.ops.nn import rnn_cell
+from TFNetwork import LayerNotFound
```
Comment (patrick-wilken) [inline TFNetworkRecLayer.py]: I imported it here, as I use it twice now.

File: tests/test_TFEngine.py [hunk 7]
```diff
@@ -1020,6 +1020,149 @@ def make_net_dict(task):
...
+  data.init_seq_order()
+
+  dec_time = DimensionTag(kind=DimensionTag.Types.Spatial, description="dec time")
+
+  config = Config()
+  config.update({
+    "model": "%s/model" % _get_tmp_dir(),
+    "batch_size": 5000,
+    "max_seqs": 2,
+    "extern_data": {"data": {"dim": n_data_dim, "sparse": False},
```
Comment (patrick-wilken) [inline tests/test_TFEngine.py]: (Ah, works again)
This should be an issue on it's own. When using MetaDataset I have to mark the time dims as equal, also for in real configs. Is this expected?

File: TFNetworkRecLayer.py [hunk 8]
```diff
@@ -2431,11 +2433,56 @@ def get_choice_seq(choice_base):
...
+
+        choice_beams = initial_beam_choices
+        if choices_seq_until_end_layer:  # otherwise end layer already corresponds to latest choice
+          for choice_ in choices_seq_until_end_layer:
+            assert_min_tf_version((1, 1), "gather_nd")
+            idxs_exp = nd_indices(choice_beams)  # (batch, beam_out, 2) -> (batch idx, beam idx)
+            src_choice_beams = self.final_acc_tas_dict["choice_%s" % choice_.name].read(
+              -1, name="ta_read_choice")  # (batch, beam) -> beam_in idx
+            assert src_choice_beams.get_shape().ndims == 2
+            choice_beams = tf.gather_nd(src_choice_beams, idxs_exp)  # (batch, beam_out)
```
Comment (albertz) [inline TFNetworkRecLayer.py]: I don't think it's a good idea to replicate the logic of `SelectSearchSourcesLayer`. Although it might be a bit tricky to resolve that in an easy and clean way (without making it too abstract/complicated).

I also don't think it's a good idea to use `tf.gather_nd` and `nd_indices` directly, and replicate the logic of `select_src_beams`.

File: TFNetworkRecLayer.py (hunk 4 above)
Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: Ah, it cannot be None. It's None in `get_output` if there are no layers in the loop, but then we would never get here.

File: TFNetworkRecLayer.py [hunk 9]
```diff
@@ -2350,33 +2351,34 @@ def cond(i, net_vars, acc_tas, seq_len_info=None):
...
     which is determined by the latest search choice given by the layer.
     This assumes that we also have `self.final_acc_tas_dict["choice_%s" % choice_base.name]` available.
+    In addition, we resolve the sequence lengths (whose beams correspond to the end layer) to the final search choices.
 
     :param str layer_name:
     :param tf.TensorArray acc_ta: accumulated outputs of that layer
     :param final_net_vars:
-    :return: (new acc_ta, final search choices)
-    :rtype: (tf.TensorArray,SearchChoices|None)
+    :param tf.Tensor|None seq_len: shape (batch * beam,)
```
[the remark this answers is not in the document]
Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: There are two cases: If `have_known_seq_len`, we get the beam from the "output" layer, otherwise from the "end" layer. I will document it.

File: TFNetworkRecLayer.py (hunk 8 above)
Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: I just made it work using `select_src_beams` (this could also be done in `transform()`). But using `SelectSearchSourcesLayer` is not easily possible, because we need to get the search choices from the TensorArray.

File: tests/test_TFEngine.py (hunk 7 above)
Comment (albertz) [reply tests/test_TFEngine.py]: Yes, this is expected. How/why should it assume that they are the same?

File: TFNetworkRecLayer.py [hunk 10]
```diff
@@ -2834,19 +2879,20 @@ def get_loop_acc_layer(name):
       with tf.name_scope(self.layer_data_templates[name].layer_class_type.cls_get_tf_scope_name(name)):
         inner_layer = self.net.get_layer(name)
         acc_ta = loop_accumulated["output_%s" % name]
-        acc_ta, search_choices = self._opt_search_resolve(
-          layer_name=name, acc_ta=acc_ta, final_net_vars=final_net_vars)
+        acc_ta, search_choices, resolved_seq_len = self._opt_search_resolve(
+          layer_name=name, acc_ta=acc_ta, final_net_vars=final_net_vars, seq_len=seq_len)
         output = self.layer_data_templates[name].output.copy_template_adding_time_dim(time_dim_axis=0)
         output.beam = search_choices.get_beam_info() if search_choices else None
+        max_len = tf.reduce_max(resolved_seq_len) if resolved_seq_len is not None else None
```
Comment (albertz) [inline TFNetworkRecLayer.py]: You said it cannot be `None`. Then this does not make sense.

File: TFNetworkRecLayer.py (hunk 8 above)
Comment (albertz) [reply TFNetworkRecLayer.py]: Where? I still see `tf.gather_nd` in that code.

Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: Now it's pushed.

File: TFNetworkRecLayer.py (hunk 10 above)
Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: Fixed.

File: tests/test_TFEngine.py (hunk 7 above)
Comment (patrick-wilken) [reply tests/test_TFEngine.py]: So the idea is that each data key has its own time dim unless configured otherwise, right? But it's quite common that multiple data keys are on the same time dim (at least for me ;) ). This should be easier to configure, without importing DimensionTag into the config.

File: TFNetworkRecLayer.py (hunk 3 above)
Comment (albertz) [reply TFNetworkRecLayer.py]: Still wrong.

File: TFNetworkRecLayer.py [hunk 11]
```diff
@@ -2431,11 +2434,45 @@ def get_choice_seq(choice_base):
...
+        src_choice_beams = self.final_acc_tas_dict["choice_%s" % choice_.name].read(
+          max_seq_len - 1, name="ta_read_choice")  # (batch, beam) -> beam_in idx
+        seq_len = select_src_beams(seq_len, src_choice_beams)
+    else:
+      # Here we don't need to resolve anything, as the sequence length is the same for all hyps in the beam.
+      # However, beam size for the current output may be different from the "output" layer.
+      # Therefore take the first len in the beam and tile it to the desired beam size.
+
+      # Separate batch and beam dims
+      seq_len_beam_size = end_layer.output.beam.beam_size if end_layer else rec_layer.output.beam.beam_size
```
Comment (albertz) [inline TFNetworkRecLayer.py]: The whole branch is for `not end_layer`, so I don't understand the `if` here.

File: TFNetworkRecLayer.py (hunk 3 above)
Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: I moved the import to line 11, because I need it here and in line 2439. I can also change it back to local imports, but it's working.

File: TFNetworkRecLayer.py (hunk 11 above)
Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: Yep, copy-paste error, fixed it.

File: TFNetworkRecLayer.py [hunk 12]
```diff
@@ -2834,21 +2868,22 @@ def get_loop_acc_layer(name):
...
           if self.parent_rec_layer.output.beam:
             # seq_len should have already a beam, same as rec_layer.output
             assert output.beam == self.parent_rec_layer.output.beam
           else:
-            output.size_placeholder[0] = tile_transposed(seq_len, axis=0, multiples=output.beam.beam_size)
-            if time_dim_tag:
-              time_dim_tag.set_tag_on_size_tensor(output.size_placeholder[0])
+            output.size_placeholder[0] = tile_transposed(resolved_seq_len, axis=0, multiples=output.beam.beam_size)
+        if time_dim_tag:
+          time_dim_tag.set_tag_on_size_tensor(output.size_placeholder[0])
```
Comment (patrick-wilken) [inline TFNetworkRecLayer.py]: I was missing this change to set the tag again after a new seq_len tensor is created. Now tests are passing.

File: TFNetworkRecLayer.py (hunk 3 above)
Comment (albertz) [reply TFNetworkRecLayer.py]: I'm not speaking about the `import`. I mean exactly the line where I put the comment. The `return` is wrong.

File: TFNetworkRecLayer.py [hunk 13]
```diff
@@ -2834,21 +2868,22 @@ def get_loop_acc_layer(name):
...
+        output.size_placeholder = {0: resolved_seq_len}
         if output.beam:
           if self.parent_rec_layer.output.beam:
             # seq_len should have already a beam, same as rec_layer.output
             assert output.beam == self.parent_rec_layer.output.beam
           else:
-            output.size_placeholder[0] = tile_transposed(seq_len, axis=0, multiples=output.beam.beam_size)
-            if time_dim_tag:
-              time_dim_tag.set_tag_on_size_tensor(output.size_placeholder[0])
+            output.size_placeholder[0] = tile_transposed(resolved_seq_len, axis=0, multiples=output.beam.beam_size)
```
Comment (albertz) [inline TFNetworkRecLayer.py]: Does this make sense? If you have logic for resolving `seq_len` in `_opt_search_resolve`, shouldn't it be tiled correctly already?
In any case, the logic should not be split; `_opt_search_resolve` should correctly resolve `seq_len`.

File: TFNetworkRecLayer.py (hunk 3 above)
Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: Ahh, fixed.

File: TFNetworkRecLayer.py (hunk 13 above)
Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: I think you're right, it is not needed anymore. The assert
```
assert output.beam == self.parent_rec_layer.output.beam
```
is definitively wrong now for the case that different outputs have different latest_choices and therefore maybe different beam sizes.

As far as I understand the case distinction added in e442199, it is for the case that there is a beam that is decided inside the rec layer, right? I think with the seq_lens of each output having the beam of its latest_search_choice this case should be handled now correctly implicitly.

**Comment by albertz** [comment]
Sorry, I made too much changes in the meantime, and there is a small conflict now. Can you rebase? Thanks!

**Comment by patrick-wilken** [comment]
I tried to rebase, but please double check.

File: TFNetworkRecLayer.py [hunk 14]
```diff
@@ -2838,33 +2872,23 @@ def get_loop_acc_layer(name):
       with tf.name_scope(self.layer_data_templates[name].layer_class_type.cls_get_tf_scope_name(name)):
         inner_layer = self.net.get_layer(name)
         acc_ta = loop_accumulated["output_%s" % name]
-        if inner_layer.get_search_choices() is None:
-          search_choices = None
-        else:
-          acc_ta, search_choices = self._opt_search_resolve(
-            layer_name=name, acc_ta=acc_ta, final_net_vars=final_net_vars)
+        acc_ta, search_choices, resolved_seq_len = self._opt_search_resolve(
+          layer_name=name, acc_ta=acc_ta, final_net_vars=final_net_vars, seq_len=seq_len)
```
Comment (patrick-wilken) [inline TFNetworkRecLayer.py]: Why did you add the if statement here? Shouldn't `_opt_search_resolve` return `search_choices =None` and the original `acc_ta` in that case?

File: TFNetworkRecLayer.py [hunk 15]
```diff
@@ -2435,11 +2437,45 @@ def get_choice_seq(choice_base):
...
+      # Here we don't need to resolve anything, as the sequence length is the same for all hyps in the beam.
+      # However, beam size for the current output may be different from the "output" layer.
+      # Therefore take the first len in the beam and tile it to the desired beam size.
+
+      # Separate batch and beam dims
+      seq_len_beam_size = rec_layer.output.beam.beam_size
+      seq_len = tf.reshape(seq_len, [batch_dim, seq_len_beam_size], name="split_batch_beam")
+
+      seq_len = seq_len[:, 0:1]
+      seq_len = tf.tile(seq_len, [1, latest_beam_size], name="resize_seq_len_beam")
```
Comment (patrick-wilken) [inline TFNetworkRecLayer.py]: I also updated this. Tiling was incorrect before, luckily @talkhouli ran into this ':D

File: TFNetworkRecLayer.py (hunk 14 above)
Comment (albertz) [reply TFNetworkRecLayer.py]: No. In `_opt_search_resolve`, we use `layer_choice = self.net.get_search_choices(src=layer)`, which might not be the same, as I derived `LayerBase.get_search_choices()` for some layers. Although I think I undid that later. But maybe we can just use `layer_choice = layer. get_search_choices()` inside `_opt_search_resolve`.

File: TFNetworkRecLayer.py [hunk 16]
```diff
@@ -2352,35 +2352,38 @@ def cond(i, net_vars, acc_tas, seq_len_info=None):
...
-      return acc_ta, None
-    if layer_choice.search_choices.keep_raw:
-      return acc_ta, layer_choice.search_choices
+      return acc_ta, None, seq_len
+    search_choices = layer.get_search_choices()
+    if not search_choices:
+      return acc_ta, None, seq_len
+    if search_choices.keep_raw:
+      return acc_ta, search_choices, seq_len
+    layer_choice = search_choices.owner
```
Comment (patrick-wilken) [inline TFNetworkRecLayer.py]: Like this? I checked that search choices are the same as before for all test cases...

Comment (albertz) [reply TFNetworkRecLayer.py]: Yes exactly. Yes I think I once had a case where this was more important, but changed that later on. But still, maybe this is better.
